File: codes/DijkstraShortestReach.cpp
#include<bits/stdc++.h>
#define endl '\n'

using namespace std;

typedef pair<long long, int> pli;

int vertices, aristas, casos, origen;
long long costo;
vector<pli> grafo[200005];
long long d[200005];
int pi[200005];
set<pli> avl;

void inicializar()
{
    for(int i = 0; i <= vertices; i++)
    {
        d[i] = INT_MAX;
        pi[i] = INT_MAX;
    }

    d[origen] = 0;
    pi[origen] = -1;
}

bool relax(int a, int b, long long w)
{
    if(d[b] > d[a] + w)
    {
        d[b] = d[a] + w;
        pi[b] = a;
        return true;
    }
    else
        return false;
}

void dijkstra()
{
    inicializar();

    avl.insert(make_pair(0, origen));

    while(!avl.empty())
    {
        pli elemento = *avl.begin();
        avl.erase(avl.begin());

        for(int i = 0; i < grafo[elemento.second].size(); i++)
        {
            pli temp = grafo[elemento.second][i];

            if(relax(elemento.second, temp.second, temp.first))
            {
                set<pli>:: iterator iterador = avl.find(make_pair(d[temp.second], temp.second));
                if(iterador != avl.end())
                    avl.erase(iterador);

                avl.insert(make_pair(d[temp.second], temp.second));
            }
        }
    }
}

int main()
{
    ios_base :: sync_with_stdio(0);
    cin.tie(0);

    cin >> casos;

    while(casos--)
    {
        cin >> vertices >> aristas;

        for(int i = 0; i < aristas; i++)
        {
            int a, b;
            cin >> a >> b >> costo;

            grafo[a].push_back(make_pair(costo, b));
            grafo[b].push_back(make_pair(costo, a));
        }

        cin >> origen;
        dijkstra();

        for(int i = 1; i <= vertices; i++)
        {
            if(i != origen)
            {
                if(d[i] == INT_MAX)
                    cout << -1 << " ";
                else
                    cout << d[i] << " ";
            }
        }
        cout << endl;
    }
}

File: codes/DijkstraShortest Reach.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DijkstraShortestReach
{
    class Program
    {
        static long[] d;
        static int[] pi;

        static void inicializar(List<Tuple<long, int>>[] grafo, int origen)
        {
            d = new long[grafo.Length];
            pi = new int[grafo.Length];

            for (int i = 0; i < grafo.Length; i++)
            {
                d[i] = int.MaxValue;
                pi[i] = int.MaxValue;
            }

            d[origen] = 0;
            pi[origen] = -1;
        }

        static bool Relax(int u, int v, long w)
        {
            if (d[v] > d[u] + w)
            {
                d[v] = d[u] + w;
                return true;
            }
            else
                return false;
        }

        static void Dijkstra(List<Tuple<long, int>>[] grafo, int origen)
        {
            inicializar(grafo, origen);

            SortedSet<Tuple<long, int>> avl = new SortedSet<Tuple<long, int>>();
            avl.Add(new Tuple<long, int>(0, origen));

            while (avl.Count != 0)
            {
                Tuple<long, int> elemento = avl.Min;
                avl.Remove(avl.Min);

                for (int i = 0; i < grafo[elemento.Item2].Count; i++)
                {
                    Tuple<long, int> temp = grafo[elemento.Item2][i];

                    if (Relax(elemento.Item2, temp.Item2, temp.Item1))
                    {
                        Tuple<long, int> aux = new Tuple<long,int>(d[temp.Item2], temp.Item2);
                        if (avl.Contains(aux))
                            avl.Remove(aux);

                        avl.Add(new Tuple<long, int>(d[temp.Item2], temp.Item2));
                    }
                }
            }
        }

        static void Main(string[] args)
        {
            int casos = int.Parse(Console.ReadLine());

            while (casos > 0)
            {
                string[] linea = Console.ReadLine().Split();
                int vertices = int.Parse(linea[0]);
                int aristas = int.Parse(linea[1]);

                List<Tuple<long, int>>[] grafo = new List<Tuple<long, int>>[vertices + 1];

                for (int i = 0; i < grafo.Length; i++)
                {
                    grafo[i] = new List<Tuple<long, int>>();
                }

                for (int i = 0; i < aristas; i++)
                {
                    linea = Console.ReadLine().Split();
                    int u = int.Parse(linea[0]);
                    int v = int.Parse(linea[1]);
                    long costo = long.Parse(linea[2]);

                    grafo[u].Add(new Tuple<long, int>(costo, v));
                    grafo[v].Add(new Tuple<long, int>(costo, u));
                }

                int origen = int.Parse(Console.ReadLine());
                Dijkstra(grafo, origen);

                for (int i = 1; i <= vertices; i++)
                {
                    if (i != origen)
                    {
                        if (d[i] == int.MaxValue)
                            Console.Write("{0} " , -1);
                        else
                            Console.Write("{0} ", d[i]);
                    }
                }
                Console.WriteLine();

                casos--;
            }
        }
    }
}
